File: frontend/lib/domain/entities/gacha_config.dart
import 'card.dart';

/// Config paramétrica del gacha (precio, garantía, probabilidades, bono de
/// rareza) — editable por un superadmin vía `/api/admin/gacha-config`, no
/// vive hardcodeada en el cliente ni en el servidor.
///
/// Las probabilidades/bono se guardan como `String` (no `double`): el
/// backend los serializa como texto decimal (Pydantic `Decimal`) y así
/// evitamos perder precisión en el round-trip GET -> editar -> PUT.
class GachaPackLevelConfig {
  final int level;
  final int price;
  final int cardsPerPack;
  final CardRank? guaranteedMinRank;

  GachaPackLevelConfig({
    required this.level,
    required this.price,
    required this.cardsPerPack,
    required this.guaranteedMinRank,
  });

  factory GachaPackLevelConfig.fromJson(Map<String, dynamic> json) {
    final raw = json['guaranteed_min_rank'] as String?;
    return GachaPackLevelConfig(
      level: json['level'] as int,
      price: json['price'] as int,
      cardsPerPack: json['cards_per_pack'] as int,
      guaranteedMinRank: raw == null ? null : CardRankApi.fromApiValue(raw),
    );
  }
}

class GachaRankProbabilitiesConfig {
  final int level;
  final String hero;
  final String demigod;
  final String minorGod;
  final String majorGod;

  GachaRankProbabilitiesConfig({
    required this.level,
    required this.hero,
    required this.demigod,
    required this.minorGod,
    required this.majorGod,
  });

  factory GachaRankProbabilitiesConfig.fromJson(Map<String, dynamic> json) {
    return GachaRankProbabilitiesConfig(
      level: json['level'] as int,
      hero: json['hero'] as String,
      demigod: json['demigod'] as String,
      minorGod: json['minor_god'] as String,
      majorGod: json['major_god'] as String,
    );
  }
}

class GachaRarityProbabilitiesConfig {
  final int level;
  final String common;
  final String rare;
  final String epic;
  final String legendary;

  GachaRarityProbabilitiesConfig({
    required this.level,
    required this.common,
    required this.rare,
    required this.epic,
    required this.legendary,
  });

  factory GachaRarityProbabilitiesConfig.fromJson(Map<String, dynamic> json) {
    return GachaRarityProbabilitiesConfig(
      level: json['level'] as int,
      common: json['common'] as String,
      rare: json['rare'] as String,
      epic: json['epic'] as String,
      legendary: json['legendary'] as String,
    );
  }
}

class GachaRarityBonusConfig {
  final String common;
  final String rare;
  final String epic;
  final String legendary;

  GachaRarityBonusConfig({
    required this.common,
    required this.rare,
    required this.epic,
    required this.legendary,
  });

  factory GachaRarityBonusConfig.fromJson(Map<String, dynamic> json) {
    return GachaRarityBonusConfig(
      common: json['common'] as String,
      rare: json['rare'] as String,
      epic: json['epic'] as String,
      legendary: json['legendary'] as String,
    );
  }
}

class GachaConfigEntity {
  final List<GachaPackLevelConfig> packLevels;
  final List<GachaRankProbabilitiesConfig> rankProbabilities;
  final List<GachaRarityProbabilitiesConfig> rarityProbabilities;
  final GachaRarityBonusConfig rarityBonus;

  GachaConfigEntity({
    required this.packLevels,
    required this.rankProbabilities,
    required this.rarityProbabilities,
    required this.rarityBonus,
  });

  factory GachaConfigEntity.fromJson(Map<String, dynamic> json) {
    return GachaConfigEntity(
      packLevels: (json['pack_levels'] as List)
          .map((e) => GachaPackLevelConfig.fromJson(e as Map<String, dynamic>))
          .toList(),
      rankProbabilities: (json['rank_probabilities'] as List)
          .map((e) => GachaRankProbabilitiesConfig.fromJson(e as Map<String, dynamic>))
          .toList(),
      rarityProbabilities: (json['rarity_probabilities'] as List)
          .map((e) => GachaRarityProbabilitiesConfig.fromJson(e as Map<String, dynamic>))
          .toList(),
      rarityBonus: GachaRarityBonusConfig.fromJson(
        json['rarity_bonus'] as Map<String, dynamic>,
      ),
    );
  }
}

File: backend/app/schemas/gacha_config.py
from decimal import Decimal
from typing import Optional

from pydantic import BaseModel, ConfigDict

from app.models.enums import Rank, Rarity


class PackLevelOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    level: int
    price: int
    guaranteed_min_rank: Optional[Rank]


class PackLevelUpdateRequest(BaseModel):
    price: int
    guaranteed_min_rank: Optional[Rank] = None


class RankProbabilitiesOut(BaseModel):
    level: int
    hero: Decimal
    demigod: Decimal
    minor_god: Decimal
    major_god: Decimal


class RankProbabilitiesUpdateRequest(BaseModel):
    hero: Decimal
    demigod: Decimal
    minor_god: Decimal
    major_god: Decimal


class RarityProbabilitiesOut(BaseModel):
    level: int
    common: Decimal
    rare: Decimal
    epic: Decimal
    legendary: Decimal


class RarityProbabilitiesUpdateRequest(BaseModel):
    common: Decimal
    rare: Decimal
    epic: Decimal
    legendary: Decimal


class RarityBonusOut(BaseModel):
    common: Decimal
    rare: Decimal
    epic: Decimal
    legendary: Decimal


class RarityBonusUpdateRequest(BaseModel):
    common: Decimal
    rare: Decimal
    epic: Decimal
    legendary: Decimal


class GachaConfigDump(BaseModel):
    pack_levels: list[PackLevelOut]
    rank_probabilities: list[RankProbabilitiesOut]
    rarity_probabilities: list[RarityProbabilitiesOut]
    rarity_bonus: RarityBonusOut


RANK_FIELDS = (Rank.hero, Rank.demigod, Rank.minor_god, Rank.major_god)
RARITY_FIELDS = (Rarity.common, Rarity.rare, Rarity.epic, Rarity.legendary)
